content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,250 @@ Like a segment tree on values rather than indices.
+and perform a binary search for the number of indices less than $𝑟$
+However, let's explore a different approach that can also be extended to the
+second type of query.
+
+### A different way
+**A different way to find the Index of $𝑟$ in the list of vertices**
+
+Instead of binary searching on the leaf, we update $𝑟$ as we recurse down the
+tree.
+If we are able to determine the position (index) of $𝑟$ in the left and right
+children of a node
+We can recurse down the tree and determine its position in the leaf node.
+
+### Determining the Position in Child Nodes
+
+To find the position of $𝑟$ in a node's left and right children, we need to
+determine how many indices are smaller than the middle value (mid) and precede
+$𝑟$.
+This can be done using a prefix sum.
+
+Let's define:
+- $c[i]$ = as 1 if $index[i]$ is smaller than mid otherwise 0
+- $prefixB[i]$ as prefix sum of $c[i]$
+
+Formally
+
+$$
+c[i] = index[i] < mid ? 1 : 0;
+prefixB[i] = prefixB[i - 1] + c[i]
+$$
+
+### Updating $𝑟$ as We Recurse
+
+To know the value of 𝑟 if we recurse left, we use prefixB[r]
+If we recurse right, we use 𝑟 - prefixB[r]
+
+## Solving the second type of query
+**Given a range l, r find the k smallest element**
+
+We will determine whether the answer for a given node is in the left or the
+right segment.
+We can calculate how many times the elements within the segments' ranges appear
+in our range $(l, r)$ using our first type of query.
+Note that this also works for non-leaf nodes using the following formula:
+
+$$
+occurrences(l, r) = r - l
+$$
+<Info title="Simular">
+This is simular to counting how many times a value appears up to index 𝑅 in our previous query. We did this by using the new 𝑅 value at the leaf node. But now, we consider the difference between the updated 𝑅 and 𝐿
+</Info>
+
+Therefore, the occurrences of the left node is
+
+$$
+left_occurrences = prefixB[r] - prefixB[l]
+$$
+
+<Info title="Left occurrences">
+  Note that left_occurrences is the amount of indices between l and r that their value is less than mid
+
+</Info>
+
+- If left_occurrences left_occurrences is greater or equal to $k$, it means the $k$-th smallest element is in
+  the left subtree. Therefore, we update our range and recurse into the left
+  child
+- If left_occurrences is less than $k$, it means the
+  $k$-th smallest element is in the right subtree. We adjust k by subtracting
+  left_occurrences from $k$, update our range, and recurse into the right child
+
+<Info title="Notice">
+  Notice we still update $l, r$ accordingly when we go left or right
+</Info>
+
+the answer then will be the value of the node we end up on (leaf)
+
+## Time complexity
+For each query, we traverse left or right, meaning the number of nodes visited is proportional to the height of the tree.
+For each node, we compute the $l$ or $r$  using prefix sums, which is $\mathcal{O}(1)$  for each query.
+Hence, the overall time complexity is $\mathcal{O}(\log n)$
```

Comment:
uh add an example problem and then we can discuss this

Reply:
yeah so if you look at the other solutions you'll see that we add the complexity below the implementation header with just a
```md
**Time Complexity:** $\mathcal{O}(?)$
```